### PR TITLE
[4/n] Stabilize GCS/Autoscaler interface: Expose placement group demands

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR adds placement group demands to the scheduling status for the new monitor service.

This is the second to last PR for migrating scheduling. (The final one will be for `autoscaler.sdk.request_resources()`) 

## Related issue number

Related to https://github.com/ray-project/ray/issues/31826

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
cc @ray-project/ray-autoscaler-stakeholders

File: src/ray/gcs/gcs_server/test/gcs_monitor_server_test.cc [hunk 1]
```diff
@@ -51,30 +52,62 @@ rpc::ResourceDemand constructResourceDemand(
   return demand;
 }
 
+std::shared_ptr<gcs::GcsPlacementGroup> constructPlacementGroupDemand(
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/test/gcs_monitor_server_test.cc:55]: nit: ConstructPlacementGroupDemand

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h [hunk 2]
```diff
@@ -369,6 +370,25 @@ class GcsPlacementGroupManager : public rpc::PlacementGroupInfoHandler {
     usage_stats_client_ = usage_stats_client;
   }
 
+  /// Get a read only view of the pending placement groups.
+  ///
+  /// \return Pending placement groups.
+  const absl::btree_multimap<
+      int64_t,
+      std::pair<ExponentialBackOff, std::shared_ptr<GcsPlacementGroup>>>
+      &GetPendingPlacementGroups() const;
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.h:379]: Interesting... so we don't expose these infos before? How is the autoscaler currently working? Is this done in a push way inside placement group?

**Review by fishbone — COMMENTED** [review]
Not cover the monitor's detailed logics. Others lgtm. Have one question for you in the comment.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h (hunk 2 above)
Comment (wuisawesome) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.h:379]: Yeah right now it's weird. `GcsResourceManager` stores a `placement_group_load_` variable that `GcsPlacementGroupManager` pushes a value to periodically. 

[I saw there was a TODO not to stop doing that](https://github.com/ray-project/ray/blob/dc68ec95eaa3b16a350b0dce55485edd0c2bcc38/src/ray/gcs/gcs_server/gcs_resource_manager.h#L47), so I added this function instead.

Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.h:379]: This is much cleaner.

**Review by fishbone — APPROVED** [review]

**Comment by wuisawesome** [comment]
Test failures all look like unrelated library failures, merging

**Pull request merged**